sdl2promela/translator.py

```diff
@@ -536,6 +548,30 @@ def __generate_input_function(
                 .build()
             )
 
+        # Generate exit procedure calls in case of composite state.
+        exitlist = []
+        current = ""
+        state_tree = state.name.split(SEPARATOR)
+
+        process = context.sdl_model.source
+        while state_tree:
+            current = current + state_tree.pop(0)
+            for comp in process.composite_states:
+                if current.lower() == comp.statename.lower():
+                    if comp.exit_procedure:
+                        exitlist.append(current)
+                    process = comp
+                    current = current + SEPARATOR
+                    break
+
+        trans = context.sdl_model.transitions[input_block.transition_id]
+        for each in reversed(exitlist):
+            if trans and all(each.startswith(st) for st in trans.possible_states):
+                exit_procedure_name = (
+                    f"{context.sdl_model.process_name}{SEPARATOR}{each}{SEPARATOR}exit"
+                )
+                statements.append(CallBuilder().withTarget(exit_procedure_name).build())
+
```

Comment:
This block needs to be refactored... possibly with parts extracted. Variable names should be improved (comp, st, trans... each could probably do in a simpler loop, but here it contributes to low readibility (there are foreach/for each constructs in different languages and here the iteration variable gets visually lost). Also, current what?

Reply:
refactored